text/0000-const-trait-impls.md

```diff
@@ -0,0 +1,1006 @@
+
+When calling a trait method, this simple scheme (that works great for free functions and inherent methods) does not work.
+
+Throughout this document, we'll be revisiting the example below. Method syntax and `dyn Trait` problems all also exist with static method calls, so we'll stick with the latter to have the simplest examples possible.
+
+```rust
+const fn default<T: Default>() -> T {
+    T::default()
+}
+
+// Could also be `const fn`, but that's an orthogonal change
+fn compile_time_default<T: Default>() -> T {
+    const { T::default() }
+}
+```
+
+Neither of the above should (or do) compile.
+The first, because you could pass any type T whose default impl could
+
+* mutate a global static,
+* read from a file, or
+* just allocate memory,
+
+which are all not possible right now in const code, and some can't be done in Rust in const code at all.
+
+It should be possible to write `default` in a way that allows it to be called in const contexts
+for types whose `Default` impl's `default` method satisfies all rules that `const fn` must satisfy
+(including some annotation that guarantees this won't break by accident).
+It must always be possible to call `default` outside of const contexts with no limitations on the generic parameters that may be passed.
+
+Similarly it should be possible to write `compile_time_default` in a way that also requires calls
+outside of const contexts to only pass generic parameters whose `Default::default` method satisifies
+the usual `const fn` rules. This is necessary in order to allow a const block
+(which can access generic parameters) in the function body to invoke methods on the generic parameter.
+
+So, we need some annotation that differentiates a `T: Default` bound from one that gives us the guarantees we're looking for.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## Nomenclature and new syntax concepts
+
+### Const trait impls
+
+It is now allowed to prefix a trait name in an impl block with `const`, marking that this `impl`'s type is now allowed to
+have methods of this `impl`'s trait to be called in const contexts (if all where bounds hold, like usual, but more on this later).
+
+An example looks as follows:
+
+```rust
+impl const Trait for Type {}
+```
+
+Such impls require that the trait is a `const trait`.
+
+All method bodies in a const trait impl are [const contexts](https://doc.rust-lang.org/reference/const_eval.html#const-context).
+
+### Const traits
+
+Traits need to opt-in to being allowed to have const trait impls. Thus you need to declare your traits by prefixing the `trait` keyword with `const`:
+
+```rust
+const trait Trait {}
+```
+
+This in turn checks all methods' default bodies as if they were `const fn`, making them callable in const contexts.
```

Comment:
Since you want inline threading and I'm not sure where best to put this, but you should probably talk about how `const fn` and methods in const traits are compatible with `const Fn` traits, since you otherwise only briefly touch on `const fn` pointers, which are related but not technically the same.

Even if const closures are an unanswered question, I'd prefer if `Option::map` and similar methods were possible if you passed them `const fn`s defined explicitly.